### ISSUE TITLE
Orchard.MediaProcessing - new filter for auto-rotate and page selection

### ISSUE THREAD

**Issue by ViRuSTriNiTy** [OrchardCMS/Orchard#7810]
Hi there,

when issue #7807 gets a go i can also provide filters for auto-rotate and page selection. Both commands are in the core of ImageResizer now, so these filters should go into Orchard.MediaProcessing.

Auto-rotate filter

    public class AutoRotateFilter : Orchard.MediaProcessing.Services.IImageFilterProvider
    {
      public AutoRotateFilter()
      {
        T = Orchard.Localization.NullLocalizer.Instance;
      }

      public Orchard.Localization.Localizer T { get; set; }

      public void Describe(Orchard.MediaProcessing.Descriptors.Filter.DescribeFilterContext aDescribe)
      {
        aDescribe.For("Transform", T("Transform"), T("Transform"))
          .Element(
            "AutoRotate",
            T("Auto-rotate"),
            T("Auto-rotates according to EXIF rotation data from the camera. Make sure the filter is first in the filter chain to ensure access to the EXIF rotation data."),
            ApplyFilter,
            DisplayFilter,
            null // no configuration form needed yet
          );
      }

      public void ApplyFilter(Orchard.MediaProcessing.Descriptors.Filter.FilterContext aContext)
      {
        var lInstructions = new ImageResizer.Instructions();

        // enable auto rotate to respect EXIF orientation, http://imageresizing.net/docs/v4/plugins/autorotate
        lInstructions.Add("autorotate", "true");

        var lResult = new System.IO.MemoryStream();

        if (aContext.Media.CanSeek)
          aContext.Media.Seek(0, System.IO.SeekOrigin.Begin);

        ImageResizer.ImageBuilder.Current.Build(aContext.Media, lResult, lInstructions);

        aContext.Media = lResult;
      }

      public Orchard.Localization.LocalizedString DisplayFilter(Orchard.MediaProcessing.Descriptors.Filter.FilterContext aContext)
      {
        return T("Auto-rotate the image according to EXIF rotation data from the camera");
      }
    }

Page selection filter

    using System.Linq;
    using Dennree.Core.Extensions.EnumerableExtensions;
    using Orchard.Mvc.Html;

    internal class PageFilterConstants
    {
      public const char SupportedExtensionsSeparator = ',';
    }

    public class PageFilter : Orchard.MediaProcessing.Services.IImageFilterProvider
    {
      public PageFilter()
      {
        T = Orchard.Localization.NullLocalizer.Instance;
      }

      public Orchard.Localization.Localizer T { get; set; }

      public void Describe(Orchard.MediaProcessing.Descriptors.Filter.DescribeFilterContext aDescribe)
      {
        aDescribe.For("Selection", T("Selection"), T("Selection"))
          .Element(
            "Page",
            T("Page"),
            null, //T("Chooses a page for rendering a multi-page capable media."),
            ApplyFilter,
            DisplayFilter,
            "PageFilter"
          );
      }

      public void ApplyFilter(Orchard.MediaProcessing.Descriptors.Filter.FilterContext aContext)
      {
        var lCurrentFileExtension = System.IO.Path.GetExtension(aContext.FilePath).ToLower();
        if (lCurrentFileExtension.StartsWith("."))
          lCurrentFileExtension = lCurrentFileExtension.Remove(0, 1);

        var lSupportedFileExtensions = GetSupportedFileExtensions(aContext);

        if (lSupportedFileExtensions.Contains(lCurrentFileExtension))
        {
          var lInstructions = new ImageResizer.Instructions();

          int lPageNumber = aContext.State.PageNumber;

          lInstructions.Add("page", lPageNumber.ToString());

          var lResult = new System.IO.MemoryStream();

          if (aContext.Media.CanSeek)
            aContext.Media.Seek(0, System.IO.SeekOrigin.Begin);

          ImageResizer.ImageBuilder.Current.Build(aContext.Media, lResult, lInstructions);

          aContext.Media = lResult;
        }
      }

      public Orchard.Localization.LocalizedString DisplayFilter(Orchard.MediaProcessing.Descriptors.Filter.FilterContext aContext)
      {
        var lSupportedFileExtensions = GetSupportedFileExtensions(aContext);

        return T.Plural(
          "Choose page {1} to render a file with extension {2} to an image",
          "Choose page {1} to render files with extension {2} to an image",
          lSupportedFileExtensions.Count(),
          (int)aContext.State.PageNumber,
          lSupportedFileExtensions.Formulate(T("and").Text)
        );
      }

      private System.Collections.Generic.IEnumerable<string> GetSupportedFileExtensions(Orchard.MediaProcessing.Descriptors.Filter.FilterContext aContext)
      {
        return ((string)aContext.State.SupportedFileExtensions)
          .ToLower()
          .Split(PageFilterConstants.SupportedExtensionsSeparator)
          .Select(e => e.Trim());
      }
    }

    public class PageFilterForms : Orchard.Forms.Services.IFormProvider
    {
      // public
        public PageFilterForms(Orchard.DisplayManagement.IShapeFactory aShapeFactory)
        {
          ShapeFactory = aShapeFactory;
          T = Orchard.Localization.NullLocalizer.Instance;
        }

        public void Describe(Orchard.Forms.Services.DescribeContext aContext)
        {
          System.Func<Orchard.DisplayManagement.IShapeFactory, object> lForm =
            aShape => {
              var f = ShapeFactory.Form(
                Id: "PageFilter",
                _SupportedExtensions: ShapeFactory.Textbox(
                  Id: "supportedFileExtensions", Name: "SupportedFileExtensions",
                  Title: T("Supported file extensions"),
                  Value: string.Format("pdf{0}tif{0}tiff", PageFilterConstants.SupportedExtensionsSeparator),
                  Description:
                    T("The file extensions of the multi-page capable media that should be supported by this filter. Multiple file extensions must be separated by the {0} character.",
                      PageFilterConstants.SupportedExtensionsSeparator),
                  Classes: new[] { "text small" }
                ),
                _PageNumber: ShapeFactory.Textbox(
                  Id: "pageNumber", Name: "PageNumber",
                  Title: T("Page number"),
                  Value: 1,
                  Description: T("The number of the page to choose for rendering a multi-page capable media."),
                  Classes: new[] { "text small" }
                )
              );

              return f;
            };

          aContext.Form("PageFilter", lForm);
        }

        public Orchard.Localization.Localizer T { get; set; }

      // protected
        protected dynamic ShapeFactory { get; set; }
    }

Should i provide a PR?

**Comment by Hazzamanic** [comment]
I think we should just add the autorotate into the current resize implementation. Or at least make it an option? I think when resizing it would be reasonable to expect it to respect exif metadata. It also more performant as you only need a single pass in imageresizer. And lastly, ImageResizer strips metadata from the image so if you resize and then autorotate the autorotate will not work.

Other than that, I think this is a great and essential addition!

**Comment by ViRuSTriNiTy** [comment]
> I think we should just add the autorotate into the current resize implementation. Or at least make it an option? I think when resizing it would be reasonable to expect it to respect exif metadata. It also more performant as you only need a single pass in imageresizer.

I've recognized the multipass issue too. Perhaps this issue needs to be addressed to avoid calling `ImageBuilder.Current.Build` multiple times. My idea is to introduce *filter groups* whereas `ApplyFilter` is called on a filter group instead on each filter separately. This way all image resizer filters could be grouped and `ImageBuilder.Current.Build` could be reduced to one single call. But this would break existing setups i think.

> And lastly, ImageResizer strips metadata from the image so if you resize and then autorotate the autorotate will not work.

Yep, thats why the filter description states

    Make sure the filter is first in the filter chain to ensure access to the EXIF rotation data.

**Comment by Hazzamanic** [comment]
So it does, didn't scroll that far across on my phone. Sorry, ignore me then!

And yes it would be nice to see an implementation where you can group image resizer implementations. That's a really nice idea. Could we leave the existing implementation and build a grouped version alongside it so as not to introduce a breaking change?

**Comment by sebastienros** [comment]
I think resize should rotate the image if necessary before. And maybe still provide it as a filter if users don't want to resize but still apply the exif.

**Comment by Hazzamanic** [comment]
Agreed, especially as thumbnails use the inline syntax for resizing, not an actual profile that we can add rotate too. And we don't want sideways thumbnails!

**Comment by ViRuSTriNiTy** [comment]
@sebastienros Thanks for your input. I'm not a fan of doing auto-rotate in the resize filter as this would introduce the need for some kind of documentation about why this is done in the resize filter and the reason would be *because the ResizeMediaUrl shape needs it*. 

How about this idea

- both filters as originally suggested 
- extend implementation of `ResizeMediaUrl()` to add an additional filter record for auto-rotate
- add a `MediaUrl` shape overload that accepts an array of filter records instead of a single one (because it's called by `ResizeMediaUrl()`)

This way one could assemble image profiles as needed by picking *resize*  and / or *auto-rotate*. Also the `ResizeMediaUrl` shape would support EXIF data and it wouldn't break existing code that uses `MediaUrl`.

**Comment by sebastienros** [comment]
Quick question: when images are displayed in the Media, are they automatically rotated? If yes then it would make sense to also do it automatically when we render a thumbnail. But I would assume they are not as we already do apply a Resize filter in this window.

**Comment by ViRuSTriNiTy** [comment]
@sebastienros 

> when images are displayed in the Media, are they automatically rotated

What do you mean with Media? If you mean the management ui for the media library then no, the thumbnails displayed by the `ResizeMediaUrl` shape aren't rotated automatically. That's why i suggested:

> extend implementation of ResizeMediaUrl() to add an additional filter record for auto-rotate

**Comment by sebastienros** [comment]
Should it be a custom record, or just a default behavior?
